node/trace/span.js

```diff
@@ -0,0 +1,203 @@
+var Ready = require('ready-signal');
+var crypto = require('crypto');
+var NullLogtron = require('null-logtron');
+
+module.exports = Span;
+module.exports.Endpoint = Endpoint;
+module.exports.Annotation = Annotation;
+module.exports.BinaryAnnotation = BinaryAnnotation;
+
+function mathRng(cb) {
+    cb(null, Math.random() * 100000000000000000);
+}
+
+function cryptoRng(cb) {
+    crypto.pseudoRandomBytes(8, cb);
+}
+
+function rng(logger, cb) {
+    cryptoRng(function cryptoRngDone(err, data) {
+        if (err) {
+            logger.error('TChannel tracing: rng error', err);
+            return mathRng(function (err, data) {
+                // TODO: this is actually wrong and should store it into
+                // a buffer, assuming crypto won't fail for now during
+                // testing
+                cb(data);
+            });
+        }
+
+        cb(data);
+    });
+}
+
+function Span(options) {
+    if (!(this instanceof Span)) {
+        return new Span(options);
+    }
+    var self = this;
+
+    // TODO: options validation
+
+    self.logger = options.logger || NullLogtron();
+
+    self._idReady = Ready();
```

Comment:
building on all the above then, if we use inlined sync PRNGs, there's no need for all this ready signalling